### ISSUE TITLE
Convert me, convert me not

### ISSUE THREAD

**Issue by som-snytt** [scalapuzzlers/scalapuzzlers.github.com#121]
I answered this question, then I thought, Hm, that's a source of mild puzzlement.

The pleasant part is the symmetry of the two invocations.

It's also a twist on the endless discussions about when to have an empty param list or none.

```
class Gate {
  def unary_!(): Gate = this
}

class Foo {
  private def foo(f: () => Gate) = 1
  val gate = new Gate

  // Compiles, -Xprint:typer shows it becomes
  //   Foo.this.foo({
  //     (() => Foo.this.gate.unary_!())
  //   })
  foo(!gate)

  // Does not compile, we get
  //  error: type mismatch;
  //  found   : Gate
  //  required: () => Gate
  //  foo(gate)
  foo(gate)
}
```

http://stackoverflow.com/q/29815264/1296806

**Comment by som-snytt** [comment]
Actually, given 

https://issues.scala-lang.org/browse/SI-7187

and

https://issues.scala-lang.org/browse/SI-9178

including

>I'd be inclined to go further and deprecate eta-expansion given an expected type of Function0, and avoid triggering it with SAM types of the same shape from the outset.

**Comment by demobox** [comment]
Wow - nice! How about the following variant, which adds a bit more symmetry:

```
class Gate {
  def unary_- : Gate = this
  def unary_+(): Gate = this
}
  
def foo(f: () => Gate) = 1
val gate = new Gate

foo(+gate)
foo(-gate) // compiler barf
```
?

**Comment by demobox** [comment]
Just following the discussion in [SI-7187](https://issues.scala-lang.org/browse/SI-7187) - let's keep this open for now (unmerged) and close depending on the outcome of that discussion?

**Comment by demobox** [comment]
@som-snytt: Actually...I think the underlying reason for the last example is **not** covered by SI-7187, so is a "legitimate" puzzler: what seems to be happening there is that `-gate` is typed as `Gate` due to the first clause of [6.26.2](http://www.scala-lang.org/files/archive/nightly/pdfs/ScalaReference.pdf):

"Evaluation. A parameterless method m of type => T is always converted to type
T by evaluating the expression to which m is bound"

Only `+gate`, which is **not** a parameterless method, becomes a candidate for eta-expansion.

The question remains whether the proposed fix for SI-7187 - applying Empty Application before Eta Expansion, would cause `+gate` to _also_ fail to compile, or whether the compile would conclude that Empty Application does not produce the correct type, and then would try Eta Expansion.

In the latter case, `-gate` and `+gate` would continue to behave differently...and puzzlingly ;-)

**Comment by som-snytt** [comment]
What you say is true, so I don't know why the first bold not.

The likely fix, it sounds like, is that eta-expand won't apply for () => T. Then empty apply of unary_+() gives a type error, as does eval of unary_-.

**Comment by demobox** [comment]
> What you say is true, so I don't know why the first bold not.

@som-snytt: The bug as currently stated is that eta expansion is tried before empty application. The different in behaviour of `-gate` and `+gate` in the example should not change if that order is changed, i.e. if empty application is tried before eta expansion.

[@retronym's proposal](https://issues.scala-lang.org/browse/SI-7187?focusedCommentId=72359&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-72359) to address the more fundamental issue by removing eta expansion where the required type is `Function0` _would_ get rid of this puzzler, but that is a broader change that would (as far as I can make out) actually require a change to the language spec.

In other words, given the language spec _as it currently stands_ (whether desirable or not), the `-gate`/`+gate` difference is accurate and Not a Bug, which would make it a legitimate puzzler.

Does that make some kind of sense?

**Comment by som-snytt** [comment]
@demobox this is my 4th day of fevered flu, so nothing makes sense. But empty apply is a type error, you error out and don't keep trying stuff. The spec word is "otherwise", or in 6.26.1, "If none of the previous conversions applies" (where I noticed recently it gives the false impression that after overload resolution it won't try views). So "applies" != "typechecks".

The method cases look like they want to be exclusive of each other: `=> T, (implicit x: X) => T, (x: X) => T, () => T`.

**Comment by demobox** [comment]
> So "applies" != "typechecks".

I see what you mean. While the Eta Expansion clause includes a qualifier that it should only be tried for certain expected types, the Empty Application one doesn't, so would be applied to `+gate` regardless of the fact that the resulting type is not what's expected, indeed.

In which case, though, the statement in the issue description of "There's no upside to it. Empty application cannot hurt eta expansion, but eta expansion can hurt empty application." isn't actually correct? I.e. here we see an example where it would cause code to break?

I'll add a comment to the list, and close this one again. Thanks for clarifying, and get well soon!

**Comment by demobox** [comment]
> I'll add a comment to the issue

Comment [here](https://issues.scala-lang.org/browse/SI-7187?focusedCommentId=72383&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-72383).

**Comment by demobox** [comment]
Thinking about this again...I think this would still qualify as a puzzler at the current point in time, since the behaviour observed is definitely in accordance with the spec as it currently stands. SI-7187 is really a ticket to _change the language spec_ in a way that I think will be good, but it's **not** about fixing an implementation bug.

That said, it probably makes sense to wait a bit to see what happens with the ticket before creating a puzzler only to deprecate it pretty much immediately.

**Comment by som-snytt** [comment]
People think the puzzlers guy has it easy, but really it's a lot of hard work and soul searching.

**Comment by demobox** [comment]
> People think the puzzlers guy has it easy, but really it's a lot of hard work and soul searching.

That [Himalayan retreat](https://github.com/scalapuzzlers/scalapuzzlers.github.com/issues/120#issuecomment-87153459) sounds better and better by the ~~minute~~puzzler ;-)

**Comment by som-snytt** [comment]
Visiting the Himalayas this week would entail more hard work and soul searching than I can muster.

**Comment by demobox** [comment]
> Visiting the Himalayas this week would entail more hard work and soul searching than I can muster.

Yes, not ideal timing, indeed :-(

**Comment by demobox** [comment]
See https://github.com/scalapuzzlers/scalapuzzlers.github.com/pull/123
